xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -37,87 +37,51 @@
 import javax.annotation.concurrent.Immutable;
 
 /**
- * Loads configuration information to bootstrap xDS load balancer.
+ * Loads configuration information to bootstrap gRPC's integration of xDS protocol.
  */
-@Immutable
 abstract class Bootstrapper {
 
   private static final String BOOTSTRAP_PATH_SYS_ENV_VAR = "GRPC_XDS_BOOTSTRAP";
 
-  static Bootstrapper getInstance() throws Exception {
-    if (FileBasedBootstrapper.defaultInstance == null) {
-      throw FileBasedBootstrapper.failToBootstrapException;
-    }
-    return FileBasedBootstrapper.defaultInstance;
+  static Bootstrapper getInstance() {
+    return new FileBasedBootstrapper();
   }
 
   /**
-   * Returns the URI the traffic director to be connected to.
-   */
-  abstract String getServerUri();
-
-  /**
-   * Returns a {@link Node} message with project/network metadata in it to be included in
-   * xDS requests.
+   * Returns configurations from bootstrap.
    */
-  abstract Node getNode();
+  abstract BootstrapInfo readBootstrap() throws Exception;
 
-  /**
-   * Returns the credentials to use when communicating with the xDS server.
-   */
-  abstract List<ChannelCreds> getChannelCredentials();
-
-  @VisibleForTesting
-  static final class FileBasedBootstrapper extends Bootstrapper {
+  private static final class FileBasedBootstrapper extends Bootstrapper {
 
-    private static final Exception failToBootstrapException;
-    private static final Bootstrapper defaultInstance;
+    private static Exception failToBootstrapException;
```

Comment:
They should be non static, otherwise these mutable non-volatile static fields can be read and written by multiple threads are not synchronized correctly.

Reply:
I don't think we need to consider multi-threading case here. But we do want to ensure the file is read only once.

Reply:
This is shared by the entire JVM, which can have arbitrary number of grpc-client channels concurrently.

Reply:
Fair. Made it thread-safe.